src/compiler/transformer.ts

```diff
@@ -171,7 +210,103 @@ namespace ts {
          * Sets flags that control emit behavior of a node.
          */
         function setNodeEmitFlags<T extends Node>(node: T, flags: NodeEmitFlags) {
-            nodeEmitFlags[getNodeId(node)] = flags;
+            const options = getEmitOptions(node, /*create*/ true);
+            if (flags & NodeEmitFlags.Merge) {
+                flags = options.flags | (flags & ~NodeEmitFlags.Merge);
+            }
+
+            options.flags = flags;
+            return node;
+        }
+
+        /**
+         * Gets a custom text range to use when emitting source maps.
+         */
+        function getSourceMapRange(node: Node) {
+            let current = node;
+            while (current) {
+                const options = getEmitOptions(current);
+                if (options && options.sourceMapRange !== undefined) {
+                    return options.sourceMapRange;
+                }
+
+                current = current.original;
+            }
+
+            return node;
+        }
+
+        /**
+         * Sets a custom text range to use when emitting source maps.
+         */
+        function setSourceMapRange<T extends Node>(node: T, range: TextRange) {
+            getEmitOptions(node, /*create*/ true).sourceMapRange = range;
+            return node;
+        }
+
+        function getTokenSourceMapRanges(node: Node) {
+            let current = node;
+            while (current) {

```

Comment:
Currently these functions are a big performance bottleneck. I'm investigating this presently, but in all likelyhood these will always be separate functions.